qutip/core/expect.py

```diff
@@ -99,7 +100,12 @@ def _single_qobj_expect(oper, state):
         )
         raise ValueError(msg)
     out = _data.expect(oper.data, state.data)
-    return out.real if oper.isherm and (state.isket or state.isherm) else out
+
+    # This ensures that expect can return something that is not a number.
+    try:
+        return out.real if oper.isherm and (state.isket or state.isherm) else out
```

Comment:
```suggestion
        return out.real if oper.isherm and (state.isket or state.isherm) and hasattr(out, "real") else out
```
I would not use a `try: ... except AttributeError:` around this expression because it accesses many attributes and already contains an `if ... else` that explicitly tests for whether `out.real` is the correct thing to return.